network/src/connectivity_manager/mod.rs

```diff
@@ -394,108 +406,171 @@ where
         }
     }
 
-    async fn dial_eligible_peers<'a>(
+    fn dial_eligible_peers<'a>(
         &'a mut self,
         pending_dials: &'a mut FuturesUnordered<BoxFuture<'static, PeerId>>,
     ) {
-        let eligible = self.eligible.read().clone();
+        let to_connect = self.choose_peers_to_dial();
+        for (peer_id, peer) in to_connect {
+            self.queue_dial_peer(peer_id, peer, pending_dials);
+        }
+    }
+
+    fn choose_peers_to_dial(&mut self) -> Vec<(PeerId, DiscoveredPeer)> {
```

Comment:
we should add a doc block here on how we choose if we dial and who we dial or add more comments inline.

In general, what I expected to see was something like:

* Identify potential upstreams (that are either outbound or inbound)
* Subtract that number from max outbound, which becomes max upstreams
* Connect to at least one peer if we are not saturated

Reply:
now we have to define what is an upstream role :P

Reply:
I'm separating that out to a different PR, as it's a lot larger problem than I first thought.  What's left should provide framework for role prioritization for Mempool / Statesync